src/backend/replication/walproposer.c

```diff
@@ -72,38 +83,110 @@ CombineHotStanbyFeedbacks(HotStandbyFeedback* hs)
 	}
 }
 
+/* Initializes the internal event set, provided that it is currently null */
 static void
-ResetWalProposerEventSet(void)
+InitEventSet(void)
 {
 	if (waitEvents)
-		FreeWaitEventSet(waitEvents);
+		elog(FATAL, "double-initialization of event set");
+
 	waitEvents = CreateWaitEventSet(TopMemoryContext, 2 + n_walkeepers);
 	AddWaitEventToSet(waitEvents, WL_LATCH_SET, PGINVALID_SOCKET,
 					  MyLatch, NULL);
 	AddWaitEventToSet(waitEvents, WL_EXIT_ON_PM_DEATH, PGINVALID_SOCKET,
 					  NULL, NULL);
+}
+
+/*
+ * Updates the stored wait event for the walkeeper, given its current sockWaitState
+ *
+ * remove_if_nothing specifies whether to remove the event if the new waiting set is empty. In
+ * certain cases, we have remove_if_nothing = false because it's known that the walkeeper state will
+ * be updated immediately after if it's not waiting for any events.
+ *
+ * In general, setting remove_if_nothing = false is just an optimization; setting it to true will
+ * almost always be correct. Please leave a comment arguing for the validity of this optimization if
+ * you use it.
+ */
+static void
+UpdateEventSet(int i, bool remove_if_nothing)
+{
+	uint32 events;
+	WalKeeper* wk = &walkeeper[i];
+
+	/*
+	 * If there isn't an applicable way to update the event, we just don't bother. This function is
+	 * sometimes called when the walkeeper isn't waiting for anything, and so the best thing to do
+	 * is just nothing.
+	 */
+	if (wk->sockWaitState != WANTS_NO_WAIT)
```

Comment:
This makes sense. At first glance, properly making this change would be tricky, so I think the best thing to do is probably to figure out the best way to remove `WKSockWaitKind` entirely.

Also -- if we always require some kind of valid socket condition, then we introduce an edge case for `SS_OFFLINE`. Granted, it's a small edge case, but it's the sort of thing that I'd be concerned about introducing because it has the potential to add hard-to-reproduce bugs.

Reply:
> the best way to remove WKSockWaitKind entirely.
+1

> then we introduce an edge case for SS_OFFLINE

Which is fine, as it must never get into UpdateEventSet, no?

Reply:
Ah yeah, you're right. `UpdateEventSet` probably doesn't need anything more than an assertion that the walkeeper isn't offline then